chore: Download browsers for all platforms to release builds from single machine

kyungeunni opened the pull request:
<!-- Thanks for your pull request. Please fill all of the mandatory fields in this template. That will help us get back to you sooner. -->

<!-- If it's your first time contributing to an Elastic repo, don't forget to sign our CLA at https://www.elastic.co/contributor-agreement -->


## Summary
fixes #150
fixes #172 (fixed by 891b9f6)

<!-- What does this change do? Which issue does it solve? -->
This PR adds an additional pre-build step. When running `npm run release` script, the pre-build script will download browsers for Mac Windows, and Linux (if the browsers are already downloaded, it will skip downloading steps. The path contains the revision number so it also prevents packaging the stale browser.). Then electron builder will include the right browser for each platform as specified in `electron-builder.yml`.

<!-- If you have any screenshots or GIFs, here's where you should include them. -->


## Implementation details
I couldn't utilize the playwright's registry library, as they are hardly coupled with the host machine's OS. Instead, I've used fetchBrowser library that is used by the registry's install function. 

<!-- If you're adding a feature or implementing an enhancement, what are notable aspects about its implementation? -->


## How to validate this change
1. run`npm run pack -- -mwl`. It will digest builds in your local `dist` directory without publishing it.
2. try run each of the binaries in each platform.
  - [x] Mac
  - [x] Windows
  - [x] Linux (`.deb` for the time being)
<!-- How can others validate that what you've done is correct? -->

<!-- Is there any particular subject on which you need more input? -->

## Next Steps
- Release the binaries as described in [RELEASE.md](https://github.com/elastic/synthetics-recorder/blob/main/RELEASE.md)
- Stretch it to include Apple Silicon build
- Run the release script in CI on tags push.

apmmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Fsynthetics-recorder-mbp%2FPR-165/detail/PR-165/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Fsynthetics-recorder-mbp%2FPR-165/detail/PR-165/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Fsynthetics-recorder-mbp%2FPR-165/detail/PR-165/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Fsynthetics-recorder-mbp%2FPR-165/detail/PR-165/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://synthetics-recorder_165.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/apm-ci/transactions/view?rangeFrom=2022-02-24T15:23:11.964Z&rangeTo=2022-02-24T15:43:11.964Z&transactionName=apm-agent-rum/synthetics-recorder-mbp/PR-{number}&transactionType=job&latencyAggregationType=avg&traceId=705c55e2b848ed6f92bfb3cdecea0207&transactionId=854875cce30e3ecd)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-02-24T15:33:11.964+0000


* Duration: 16 min 1 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 95  |
| Skipped      | 0 |
| Total        | 95   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->












### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Fsynthetics-recorder-mbp%2FPR-165/detail/PR-165/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `npm build`
<ul>
<li>Took 0 min 1 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-rum/pipelines/synthetics-recorder-mbp/pipelines/PR-165/runs/1/steps/234/log/?start=0">here</a></li>
<li>Description: <code>npm run build</code></l1>
</ul>
  
  </p>
  </details>



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

justinkambic commented:
Thanks for taking this on, it represents a huge accomplishment.

In my process of smoke testing binaries I built with this branch, I encountered a problem running tests on Linux and Windows.

<img width="539" alt="image" src="https://user-images.githubusercontent.com/18429259/155589397-9cd930fd-2447-45c1-a63f-709955044c90.png">

This problem doesn't happen on the Mac build, so I think we are probably seeing the same issue here as we had with the browsers. Sharp isn't getting copied over for non-native builds, and thus it seems to be causing the test feature to fail.

vigneshshanmugam commented on electron-builder.env:
```diff
@@ -0,0 +1 @@
+BROWSER_REVISION=907428
```
We can automate this with ENV variable. We need to get it from https://github.com/elastic/synthetics-recorder/blob/13a146375d4582a04361d057969c3ab17c6e5e42/scripts/install-pw.js#L29

vigneshshanmugam commented on scripts/download-browsers.js:
```diff
@@ -0,0 +1,97 @@
+const { getChromeVersion } = require("./install-pw");
+
+const SUPPORTED_PLATFORMS = ["linux", "mac", "win"];
+const EXECUTABLE_PATHS = {
```
Need `ffmpeg` as well if i remember correctly for all platforms. Please double check if thats not the case.

kyungeunni replied:
Thanks for reminding me this!! I've updated the script to also download `ffmpeg` too.

vigneshshanmugam commented on scripts/download-browsers.js:
```diff
@@ -0,0 +1,97 @@
+  }
+}
+// eslint-disable-next-line no-console
+downloadAll().then(() => console.log("Download succeeded"));
```
Move them inside this check if we want to execute only when running as main module and not while importing - https://github.com/elastic/synthetics-recorder/blob/13a146375d4582a04361d057969c3ab17c6e5e42/scripts/install-pw.js#L50

vigneshshanmugam commented on package.json:
```diff
@@ -12,9 +12,11 @@
+    "release": "npm run pack -- -mwl --publish=always",
+    "pack": "run-s prepack electron-builder",
+    "prepack": "run-p build prepack:browsers",
+    "prepack:browsers": "node scripts/download-browsers",
```
As @justinkambic experienced, this would still break for native deps, We can fix it by running the preparation and package script serially for all platforms and uploading the pack one by one.

vigneshshanmugam commented on scripts/download-browsers.js:
```diff
@@ -0,0 +1,97 @@
+  // 'mac-arm64': '%s/builds/chromium/%s/chromium-mac-arm64.zip',
+};
+
+async function download(platform, revision, directory) {
```
Did you try to check the Registry in Playwright if there was any other way to do this?

kyungeunni replied:
Yes I went through all the Registry in Playwright and it seemed to be no work way around in there, especially I couldn't mock the platform as they would check the host machine's platform in various steps. So I used one of the functions that Registry uses to download the browser which is not tied to the host machine's platform(you pass it via parameter)

vigneshshanmugam reviewed (commented):
Great work @kyungeunni 🎉

kyungeunni commented:
This PR is ready for re-review. What I've fixed since the last review:
- Download `ffmpeg` as well as `chromium` before packing. Previously I've downloaded all the browsers for all the platform prior to packing, but I've updated it to download browsers per the requested platform. (e.g. when building binary for windows, download only for windows).
- Fixed native module download issue, namely `sharp` which is used by `@elastic/synthetics`. electron-builder supposed to handle this but it is incompatible with how sharp's install script works. more details are described [here](https://github.com/lovell/sharp/issues/2764). 

The latest build works well on Mac and Windows. When running it on Linux(Ubuntu 18.04) it had an issue launching the playback(test) but that is out of the scope as it doesn't seem to be related to the build. (I will create another ticket regarding that shortly)

cc @justinkambic @vigneshshanmugam

justinkambic commented:
Fantastic update @kyungeunni and thank you for the advice @emilioalvap ❤️ 

I will review as soon as possible.

kyungeunni commented:
Hey @lucasfcosta 👋 @justinkambic mentioned that it would be great if you could test running the build script from your machine and see if it works well just to make sure it is not customed to work only on my machine. When you have a minute, could you try running `npm run pack -- -mwl` and see all the builds are created under `/dist` without an error please? 🙏 I'm not sure

justinkambic commented:
@kyungeunni I'm still seeing an issue on Linux unfortunately 😞 

<img width="919" alt="image" src="https://user-images.githubusercontent.com/18429259/159345212-b6f6b668-f9f6-44a4-bd79-ee0ac548fae1.png">

justinkambic commented:
@kyungeunni I'm still seeing an issue on Linux unfortunately 😞 

<img width="919" alt="image" src="https://user-images.githubusercontent.com/18429259/159345212-b6f6b668-f9f6-44a4-bd79-ee0ac548fae1.png">

vigneshshanmugam commented:
Context - `.synthetics` folder gets created by Synthetics library to put the screenshots and other cached artifacts to be constructed later when the journeys are completed. We need to add permission to be able to create that folder across all OS.

kyungeunni commented:
Thanks, @vigneshshanmugam !! I'm wondering if it was a permission issue, shouldn't the error code be `EACCES: permission denied`?

lucasfcosta commented on scripts/beforePack.js:
```diff
@@ -0,0 +1,57 @@
+      stdio: "inherit",
+      shell: true,
+      env: {
+        ...process.env,
```
Here too, I'm not sure if this is a big deal, but I'd imagine that if anyone pushes a malicious update to `libvips` or any other library we use in the `fix-sharp` command, they could have an environment variable containing our secrets used for signing the app or other credentials.

Maybe I'm being paranoid or there are mechanisms to prevent this already, but I'd appreciate more thoughts from others here.

emilioalvap replied:
@lucasfcosta Good point. I'm not sure we would like to follow an allow list approach here, since we are taking into consideration multiple types (X, chromium, playwright, distro, etc...) and user-defined variables. 
IMO, it should suffice to filter those variables that are only meant to be used in the recorder itself: user auth, tokens and such.

kyungeunni replied:
Yeah, that's a valid point 👍 At the same time, I also thought would we have the same security concerns if we were to run the same script in the main process. Let me know if I'm being so naive!! Anyways It is always to be precautious about it, and I really like what @emilioalvap suggested. We should consider auditing some of the sensitive env vars and filter them out before passing to the child process.

lucasfcosta replied:
Agreed, a filter _out_ should be enough rather than a filter _in_ just to be sure we're not passing our own signing keys to a malicious package downwards.

andrewvc replied:
Just so I can better understand here, isn't there a whole stack of dependencies we already use, given that we use an OSS toolchain? I guess my question is why are we particularly concerned here? I spoke briefly with @emilioalvap , and he pointed out that NPM is perhaps weaker than a linux distro in terms of supply chain (though probably not for many attacks). 


CC @levinebw, your thoughts here?

lucasfcosta commented on scripts/download-browsers.js:
```diff
@@ -0,0 +1,135 @@
+
+exports.downloadForPlatform = async function downloadForPlatform(platform) {
+  platform = translatePlatform(platform);
+  for (const browser of ["chromium", "ffmpeg"]) {
```
Nit: should we change `browser` to `software` here and in the references within the function like `getBrowserVersion` given we're not only dealing with browsers here?

kyungeunni replied:
That's a good point, will amend 👍

lucasfcosta reviewed (commented):
Thanks for this @kyungeunni!

My only comments were around env vars and small nits.

Btw, I did build the runner on my Mac and ran the built recorder. I was able to record a multi-step journey and replay it just fine :)

vigneshshanmugam commented:
> Thanks, @vigneshshanmugam !! I'm wondering if it was a permission issue, shouldn't the error code be EACCES: permission denied? 

It seems like we are trying to access it for the screenshots, In that case, can you try with `screenshots: off` when running Synthetics test command. That should make the issue go away.

lucasfcosta commented:
> > Thanks, @vigneshshanmugam !! I'm wondering if it was a permission issue, shouldn't the error code be EACCES: permission denied?
> 
> It seems like we are trying to access it for the screenshots, In that case, can you try with `screenshots: off` when running Synthetics test command. That should make the issue go away.

@vigneshshanmugam I think even if we turn it off, the runner would still try to create a folder as specified [here](https://github.com/elastic/synthetics/blob/201f7da7db8d8ce9c13855023c8566b73ebaf91a/src/core/runner.ts#L442-L458) though, wouldn't it? Or am I missing something?

kyungeunni commented on scripts/beforePack.js:
```diff
@@ -0,0 +1,57 @@
+        reject(new Error("process finished with error code " + code));
+      }
+    });
+    npmInstall.on("error", error => {
+      reject(error);
+    });
```
Thanks updated (The less code the better 🧹)

vigneshshanmugam commented:
> @vigneshshanmugam I think even if we turn it off, the runner would still try to create a folder as specified [here](https://github.com/elastic/synthetics/blob/201f7da7db8d8ce9c13855023c8566b73ebaf91a/src/core/runner.ts#L442-L458) though, wouldn't it? Or am I missing something?

++ As you mentioned there is no way to turn it off, turning off screenshots wont just look for it.

justinkambic commented on electron/execution.js:
```diff
@@ -245,6 +245,7 @@ async function onTest(data, browserWindow) {
      */
     synthCliProcess = fork(`${SYNTHETICS_CLI}`, args, {
       env: {
+        ...process.env,
```
This is something I'm uncertain about as I'm not accustomed to the security implications of changes to CLI interaction. From speaking offline it seems like this is not a large risk, but I'd like to hear others' opinions.

kyungeunni replied:
As @emilioalvap mentioned in another comment, I think there will be various types of env values as well as platform specific ones, so it can be a bit tricky to track all of them. We should measure the effort vs worth by doing so (if we decide to do it). 

Another note, when env option is not specified when spawning, Node assigns parent's `process.env` as a [default](https://nodejs.org/api/child_process.html#child_processspawncommand-args-options)

justinkambic commented on scripts/beforePack.js:
```diff
@@ -0,0 +1,55 @@
+  return Promise.all([downloadForPlatform(platform), fixSharp(arch, platform)]);
+};
+
+const { Arch } = require("electron-builder");
```
Is there a functional reason that `Arch` needs to be required after `beforePack` is initialized?

justinkambic reviewed (commented):
I'm going to comment here and leave a passive LGTM - we should make sure that we don't leave these threads questioning security topics open before merging. I'll reach out to the other reviewers to see if they have any further thoughts on revisions we can make here, or if they're alright to merge as-is.

lucasfcosta reviewed (approved):
I'll follow @justinkambic's on this and leave this as a passive LGTM considering that filtering out the environment variables with our keys should be enough to prevent it from being passed downwards to other packages, although I do think we should just consult with security to be sure before merging.

As a note, I've also tested E2E the build for my MacOS and it worked fine for recording, replaying, and generating code for a journey.

andrewvc commented:
Works fine on windows!

vigneshshanmugam commented on scripts/afterPack.js:
```diff
@@ -0,0 +1,52 @@
+  const platform = ctx.electronPlatformName;
+  if (platform === "linux") {
+    const resourcesPath = path.join(ctx.appOutDir, "resources");
+    changePermission(resourcesPath);
```
Did you miss the return statement here?

vigneshshanmugam reviewed (approved):
LGTM, Great work @kyungeunni

vigneshshanmugam commented on scripts/beforePack.js:
```diff
@@ -0,0 +1,54 @@
+exports.default = function beforePack(ctx) {
+  const arch = Arch[ctx.arch];
+  const platform = ctx.electronPlatformName;
+  return Promise.all([downloadForPlatform(platform), fixSharp(arch, platform)]);
```
If the error handling is taken care of the electron-builder, then all good 👍🏽 I just wanted us to avoid UnhandledPromiseRejection errors.

kyungeunni replied:
<img width="913" alt="image" src="https://user-images.githubusercontent.com/16660866/161099382-1dc16f8f-f8e4-4831-b7a5-1c32b3557415.png">
yeah it's handled by electron-builder and looks like this 👍 thanks for double checking it with me!!

lucasfcosta commented:
Just adding an approval here considering the conversation we've just had with someone representing the security team (meaning we won't do any signing at this stage and thus not pass credentials down through `env`).
